hydra-node/src/Hydra/Chain.hs

```diff
@@ -114,46 +114,86 @@ data PostTxError tx
     PlutusValidationFailed {plutusFailure :: Text, plutusDebugInfo :: Text}
   | NoSeedInput
   | NoPaymentInput
-  | InvalidStateToPost {txTried :: PostChainTx tx}
+  | InvalidStateToPost {txTried :: PostChainTx tx, chainState :: ChainStateType tx}
   | UnsupportedLegacyOutput {byronAddress :: Address ByronAddr}
-  deriving (Exception, Generic, ToJSON, FromJSON)
+  deriving (Generic)
+
+deriving instance (IsTx tx, IsChainState (ChainStateType tx)) => Eq (PostTxError tx)
+deriving instance (IsTx tx, IsChainState (ChainStateType tx)) => Show (PostTxError tx)
+deriving instance (IsTx tx, IsChainState (ChainStateType tx)) => ToJSON (PostTxError tx)
+deriving instance (IsTx tx, IsChainState (ChainStateType tx)) => FromJSON (PostTxError tx)
+
+instance (IsTx tx, IsChainState (ChainStateType tx)) => Exception (PostTxError tx)
+
+instance (IsTx tx, Arbitrary (ChainStateType tx)) => Arbitrary (PostTxError tx) where
+  arbitrary = genericArbitrary
+
+-- | Types of what to keep as L1 chain state.
+type family ChainStateType tx
+
+-- | Interface available from a chain state. Expected to be instantiated by all
+-- 'ChainStateType tx'.
+class (Eq a, Show a, Arbitrary a, ToJSON a, FromJSON a) => IsChainState a where
+  -- | Get the chain slot for a chain state. NOTE: For any sequence of 'a'
+  -- encountered, we assume monotonically increasing slots.
+  chainStateSlot :: a -> ChainSlot
 
-deriving instance IsTx tx => Eq (PostTxError tx)
-deriving instance IsTx tx => Show (PostTxError tx)
+  -- TODO: document
+  advanceSlot :: a -> a
```

Comment:
This could/should be moved to a test-specific type class for chain states as it is only required in the `simulatedChainAndNetwork` of the `BehaviorSpec` (Alternative: make that `SimpleTx` specific.)